pkg/storage/stores/shipper/indexshipper/gatewayclient/gateway_client.go

```diff
@@ -365,6 +474,14 @@ func (s *GatewayClient) clientDoQueries(ctx context.Context, gatewayQueries []*l
 // poolDo executes the given function for each Index Gateway instance in the ring mapping to the correct tenant in the index.
 // In case of callback failure, we'll try another member of the ring for that tenant ID.
 func (s *GatewayClient) poolDo(ctx context.Context, callback func(client logproto.IndexGatewayClient) error) error {
+	return s.poolDoWithStrategy(ctx, callback, func(error) bool { return true })
+}
+
+func (s *GatewayClient) poolDoWithStrategy(
```

Comment:
idea: What about using this [retry pkg](https://github.com/avast/retry-go)? You can catch errors to avoid retrying, has exponential backoff, logging and context.